p2p/host/relay/autorelay.go

```diff
@@ -124,28 +136,43 @@ func (ar *AutoRelay) background(ctx context.Context) {
 	}
 }
 
-func (ar *AutoRelay) findRelays(ctx context.Context) bool {
-	if ar.numRelays() >= DesiredRelays {
-		return false
-	}
+func (ar *AutoRelay) findRelays(ctx context.Context, reachabilityChanged <-chan struct{}, updatec chan<- bool) {
+	running := false
 
-	update := false
-	for retry := 0; retry < 5; retry++ {
-		if retry > 0 {
-			log.Debug("no relays connected; retrying in 30s")
-			select {
-			case <-time.After(30 * time.Second):
-			case <-ctx.Done():
-				return update
+	for {
+		select {
+		case <-reachabilityChanged:
+			ar.mx.Lock()
+			if ar.status == network.ReachabilityPrivate {
+				running = true
+			} else if ar.status == network.ReachabilityPublic {
+				running = false
 			}
-		}
+			ar.mx.Unlock()
 
-		update = ar.findRelaysOnce(ctx) || update
-		if ar.numRelays() > 0 {
-			return update
+			if running {
+				ar.checkUpdate(ctx, updatec)
+			}
+		case <-time.After(30 * time.Second):
```

Comment:
This will never give up, even if there are no relays to find. This should probably use a bounded exponential backoff.

Reply:
@Stebalien  Thanks for your correction and I commit a bounded exponential backoff for checkUpdate.